File: tasks/TimeSeries-Downsampler/steps/milestone_1/tests/test_m1.py
"""Tests for Milestone 1: basic event parsing and tumbling window aggregation."""
import os
import shutil
import subprocess
import tempfile

_MILESTONE = "m1"


def _run_rust_verifier():
    """Run the milestone's Rust assertions from a verifier-owned cargo harness.

    The harness is created in a temporary directory outside /app and depends on
    the agent's crate via a path dependency, so no verifier files are ever
    written into the agent's workspace. Returns the set of passing test names.
    """
    test_rs = os.path.join(
        os.path.dirname(os.path.abspath(__file__)), f"test_{_MILESTONE}.rs"
    )
    harness = tempfile.mkdtemp(prefix=f"verifier_{_MILESTONE}_")
    try:
        os.makedirs(os.path.join(harness, "src"), exist_ok=True)
        os.makedirs(os.path.join(harness, "tests"), exist_ok=True)
        with open(os.path.join(harness, "Cargo.toml"), "w") as fh:
            fh.write(
                "[package]\n"
                f'name = "verifier_{_MILESTONE}"\n'
                'version = "0.0.0"\n'
                'edition = "2021"\n\n'
                "[lib]\n"
                'path = "src/lib.rs"\n\n'
                "[dependencies]\n"
                'timeseries_downsampler = { path = "/app/app" }\n\n'
                "[[test]]\n"
                f'name = "test_{_MILESTONE}"\n'
                f'path = "tests/test_{_MILESTONE}.rs"\n'
            )
        open(os.path.join(harness, "src", "lib.rs"), "w").close()
        shutil.copyfile(
            test_rs, os.path.join(harness, "tests", f"test_{_MILESTONE}.rs")
        )
        result = subprocess.run(
            ["cargo", "test",
             "--manifest-path", os.path.join(harness, "Cargo.toml"),
             "--test", f"test_{_MILESTONE}", "--offline",
             "--", "--test-threads=1"],
            capture_output=True, text=True, timeout=300,
        )
        output = result.stdout + result.stderr
        passed = set()
        for line in output.splitlines():
            line = line.strip()
            parts = line.split(" ... ")
            if line.startswith("test ") and len(parts) == 2 and parts[1].strip() == "ok":
                passed.add(parts[0][5:])
        return passed
    except Exception:
        return set()
    finally:
        shutil.rmtree(harness, ignore_errors=True)


_PASSED = _run_rust_verifier()

class TestMilestone1:
    def test_parse_valid_event(self):
        """Valid pipe-delimited strings are parsed into Event structs correctly."""
        assert "parse_valid_event" in _PASSED

    def test_parse_b64_event(self):
        """Base64 encoded IEEE 754 double precision floats are parsed correctly."""
        assert "parse_b64_event" in _PASSED

    def test_parse_valid_event_with_quotes(self):
        """Metrics with quoted names are parsed correctly, ignoring embedded pipes."""
        assert "parse_valid_event_with_quotes" in _PASSED

    def test_parse_invalid_events_returns_none(self):
        """Blank lines or lines with invalid data return None."""
        assert "parse_invalid_events_returns_none" in _PASSED

    def test_tumbling_window_calculates_correct_bounds(self):
        """Tumbling window aggregator groups events into correct time boundaries."""
        assert "tumbling_window_calculates_correct_bounds" in _PASSED

    def test_tumbling_window_aggregates_stats(self):
        """Tumbling window correctly calculates min, max, avg, count, and median."""
        assert "tumbling_window_aggregates_stats" in _PASSED

    def test_flush_window_removes_data(self):
        """Flushing a window returns its results and clears it from the aggregator's memory."""
        assert "flush_window_removes_data" in _PASSED

File: tasks/TimeSeries-Downsampler/steps/milestone_1/tests/test_m1.rs
use timeseries_downsampler::models::Event;
use timeseries_downsampler::parser::parse_event;
use timeseries_downsampler::tumbling::TumblingWindowAggregator;

#[test]
fn parse_valid_event() {
    let event = parse_event("1700000000000|evt1|cpu|45.5|parent_evt").unwrap();
    assert_eq!(event.timestamp_ms, 1700000000000);
    assert_eq!(event.event_id, "evt1");
    assert_eq!(event.name, "cpu");
    assert_eq!(event.value, 45.5);
    assert_eq!(event.dependency_ids, vec!["parent_evt".to_string()]);
    
    // Without dependency
    let event2 = parse_event("1700000000000|evt1|cpu|45.5|").unwrap();
    assert!(event2.dependency_ids.is_empty());

    // Multiple dependencies
    let event3 = parse_event("1700000000000|evt1|cpu|45.5|dep1,dep2").unwrap();
    assert_eq!(event3.dependency_ids, vec!["dep1".to_string(), "dep2".to_string()]);
}

#[test]
fn parse_b64_event() {
    // 45.5 in IEEE 754 f64 little endian is: 00 00 00 00 00 C0 46 40
    // b64 encoded: AAAAAADARkA=
    let event = parse_event("1700000000000|evt1|cpu|b64:AAAAAADARkA=|").unwrap();
    assert_eq!(event.value, 45.5);
}

#[test]
fn parse_valid_event_with_quotes() {
    let event = parse_event("1700000000000|evt1|\"cpu|usage\"|45.5|").unwrap();
    assert_eq!(event.timestamp_ms, 1700000000000);
    assert_eq!(event.event_id, "evt1");
    assert_eq!(event.name, "cpu|usage");
    assert_eq!(event.value, 45.5);
    assert!(event.dependency_ids.is_empty());
}

#[test]
fn parse_invalid_events_returns_none() {
    assert!(parse_event("").is_none());
    assert!(parse_event("1700|evt1|cpu").is_none());
    assert!(parse_event("invalid|evt1|cpu|45.5").is_none());
}

#[test]
fn tumbling_window_calculates_correct_bounds() {
    let mut agg = TumblingWindowAggregator::new(60000); // 1 minute
    let event = parse_event("60500|e1|mem|10.0").unwrap();
    agg.add_event(&event);
    
    let results = agg.flush_window(60000);
    assert_eq!(results.len(), 1);
    assert_eq!(results[0].window_start_ms, 60000);
    assert_eq!(results[0].window_end_ms, 120000);
}

#[test]
fn tumbling_window_aggregates_stats() {
    let mut agg = TumblingWindowAggregator::new(60000);
    agg.add_event(&parse_event("60100|e1|cpu|10.0").unwrap());
    agg.add_event(&parse_event("60200|e2|mem|50.0").unwrap());
    agg.add_event(&parse_event("60300|e3|cpu|30.0").unwrap());
    agg.add_event(&parse_event("60400|e4|cpu|20.0").unwrap());
    
    for i in 1..=20 {
        let line = format!("60500|e{}|disk|{}.0", i, i);
        agg.add_event(&parse_event(&line).unwrap());
    }
    
    let results = agg.flush_window(60000);
    assert_eq!(results.len(), 3);
    
    let cpu = results.iter().find(|r| r.name == "cpu").unwrap();
    assert_eq!(cpu.count, 3);
    assert_eq!(cpu.min, 10.0);
    assert_eq!(cpu.max, 30.0);
    assert_eq!(cpu.avg, 20.0);
    // Median for [10.0, 20.0, 30.0] -> N=3 (odd), middle is 20.0
    assert_eq!(cpu.median, 20.0);
    
    let mem = results.iter().find(|r| r.name == "mem").unwrap();
    assert_eq!(mem.count, 1);
    assert_eq!(mem.avg, 50.0);
    assert_eq!(mem.median, 50.0);

    let disk = results.iter().find(|r| r.name == "disk").unwrap();
    // disk values: 1.0 to 20.0 (N=20) (even)
    // Median is (10.0 + 11.0) / 2 = 10.5
    assert_eq!(disk.median, 10.5);
}

#[test]
fn flush_window_removes_data() {
    let mut agg = TumblingWindowAggregator::new(60000);
    agg.add_event(&parse_event("60100|e1|cpu|10.0").unwrap());
    
    let results_first = agg.flush_window(60000);
    assert_eq!(results_first.len(), 1);
    
    let results_second = agg.flush_window(60000);
    assert_eq!(results_second.len(), 0);
}
